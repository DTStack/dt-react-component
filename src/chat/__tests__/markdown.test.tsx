import React from 'react';
import { act } from 'react-dom/test-utils';
import { cleanup, render } from '@testing-library/react';

import Markdown from '../markdown';

jest.mock('remark-gfm', () => () => {});

describe('Test Chat Markdown', () => {
    beforeEach(cleanup);

    it('Match Snapshots', () => {
        expect(render(<Markdown># title</Markdown>).asFragment()).toMatchSnapshot('default');
        expect(
            render(
                <Markdown typing className="test">
                    {`\`inline code test \`

                    \`\`\`sql
                    SELECT * FROM table;
                    \`\`\`
                    `}
                </Markdown>
            ).asFragment()
        ).toMatchSnapshot('typing');
    });

    it('Should call onMount', () => {
        const fakeOnMount = jest.fn();
        render(
            <Markdown typing={false} onMount={fakeOnMount}>
                # title
            </Markdown>
        );
        expect(fakeOnMount).toBeCalled();
    });

    it('Should be a memoized component', () => {
        function getContainer(container: HTMLElement) {
            return container.querySelector('.dtc__aigc__markdown');
        }
        function isBlink(container: HTMLElement) {
            return !!getContainer(container)?.classList.contains('dtc__aigc__markdown--blink');
        }

        const { container, rerender } = render(<Markdown typing={false}># title</Markdown>);

        // Typing changed caused re-render
        expect(isBlink(container)).toBeFalsy();
        act(() => {
            rerender(<Markdown typing># title</Markdown>);
        });
        expect(isBlink(container)).toBeTruthy();

        // className changed caused re-render
        expect(getContainer(container)?.classList.contains('jest')).toBeFalsy();
        act(() => {
            rerender(
                <Markdown typing className="jest">
                    # title
                </Markdown>
            );
        });
        expect(getContainer(container)?.classList.contains('jest')).toBeTruthy();

        // children changed caused re-render
        expect(getContainer(container)?.firstChild?.textContent).toBe('title');
        act(() => {
            rerender(
                <Markdown typing className="jest">
                    {'```sql\nSELECT * FROM table;\n```'}
                </Markdown>
            );
        });
        expect(getContainer(container)).toMatchSnapshot('code block');

        const codeBlock = { convert: true };
        // codeBlock properties changed caused re-render
        act(() => {
            rerender(
                <Markdown typing className="jest" codeBlock={codeBlock}>
                    {'```sql\nSELECT * FROM table;\n```'}
                </Markdown>
            );
        });
        expect(getContainer(container)).toMatchSnapshot('code block with convert');

        // others changed not caused re-render
        act(() => [
            rerender(
                <Markdown
                    typing
                    className="jest"
                    codeBlock={codeBlock}
                    components={{ pre: ({ children }) => <div data-testid="pre">{children}</div> }}
                >
                    {'```sql\nSELECT * FROM table;\n```'}
                </Markdown>
            ),
        ]);
        expect(getContainer(container)?.querySelector('[data-testid="pre"]')).toBeFalsy();
    });
});
