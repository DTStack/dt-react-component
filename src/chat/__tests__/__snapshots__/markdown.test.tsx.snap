// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Chat Markdown Match Snapshots: default 1`] = `
<DocumentFragment>
  <div
    class="dtc__aigc__markdown"
  >
    <h1>
      title
    </h1>
    

    <hr
      class="dtc__aigc__markdown__hr"
      color="#ebecf0"
    />
  </div>
</DocumentFragment>
`;

exports[`Test Chat Markdown Match Snapshots: typing 1`] = `
<DocumentFragment>
  <div
    class="dtc__aigc__markdown dtc__aigc__markdown--blink test"
  >
    <div>
      <code
        class="dtc__aigc__markdown__inlineCode"
      >
        inline code test 
      </code>
    </div>
    

    <div
      class="dtc__aigc__codeblock"
    >
      <div
        class="dtc__aigc__codeblock__header"
      >
        <span
          class="dtc__aigc__codeblock__language"
        >
          sql
        </span>
        <div
          class="dtc__aigc__codeblock__tool"
        >
          <span
            class="dtc-copy"
          >
            <span
              class="dtc-copy__default-icon"
              data-mock-icon="CopyOutlined"
            />
          </span>
        </div>
      </div>
      <pre
        style="background: rgb(255, 255, 255); color: rgb(56, 56, 56); font-family: "Fira Code", "Fira Mono", Menlo, Consolas, "DejaVu Sans Mono", monospace; direction: ltr; text-align: left; white-space: pre; word-spacing: normal; word-break: normal; line-height: 1.5; tab-size: 2; hyphens: none; padding: 1em; margin: 0px; overflow: auto; border-radius: 0 0 4px 4px; max-height: 270px;"
      >
        <code
          class="language-sql"
          style="background: rgb(255, 255, 255); color: rgb(56, 56, 56); font-family: "Fira Code", "Fira Mono", Menlo, Consolas, "DejaVu Sans Mono", monospace; direction: ltr; text-align: left; white-space: pre; word-spacing: normal; word-break: normal; line-height: 1.5; tab-size: 2; hyphens: none;"
        >
          <span>
            <span
              class="linenumber react-syntax-highlighter-line-number"
              style="display: inline-block; min-width: 1.25em; padding-right: 1em; text-align: right; user-select: none; font-style: italic; color: rgb(160, 160, 160);"
            >
              1
            </span>
            <span>
                              
            </span>
            <span
              class="token identifier"
              style="color: rgb(56, 56, 56);"
            >
              \`
            </span>
            <span
              class="token identifier"
              style="color: rgb(56, 56, 56);"
            >
              \`
            </span>
            <span
              class="token identifier"
              style="color: rgb(56, 56, 56);"
            >
              \`
            </span>
            <span
              class="token identifier"
            >
              sql

            </span>
          </span>
          <span>
            <span
              class="linenumber react-syntax-highlighter-line-number"
              style="display: inline-block; min-width: 1.25em; padding-right: 1em; text-align: right; user-select: none; font-style: italic; color: rgb(160, 160, 160);"
            >
              2
            </span>
                            SELECT * FROM table;

          </span>
          <span>
            <span
              class="linenumber react-syntax-highlighter-line-number"
              style="display: inline-block; min-width: 1.25em; padding-right: 1em; text-align: right; user-select: none; font-style: italic; color: rgb(160, 160, 160);"
            >
              3
            </span>
            <span
              class="token identifier"
            >
                              \`\`
            </span>
            <span
              class="token identifier"
              style="color: rgb(56, 56, 56);"
            >
              \`
            </span>
            <span>
              

            </span>
          </span>
          <span>
            <span
              class="linenumber react-syntax-highlighter-line-number"
              style="display: inline-block; min-width: 1.25em; padding-right: 1em; text-align: right; user-select: none; font-style: italic; color: rgb(160, 160, 160);"
            >
              4
            </span>
                            
          </span>
        </code>
      </pre>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Test Chat Markdown Should be a memoized component: code block 1`] = `
<div
  class="dtc__aigc__markdown dtc__aigc__markdown--blink jest"
>
  <div
    class="dtc__aigc__codeblock"
  >
    <div
      class="dtc__aigc__codeblock__header"
    >
      <span
        class="dtc__aigc__codeblock__language"
      >
        sql
      </span>
      <div
        class="dtc__aigc__codeblock__tool"
      >
        <span
          class="dtc-copy"
        >
          <span
            class="dtc-copy__default-icon"
            data-mock-icon="CopyOutlined"
          />
        </span>
      </div>
    </div>
    <pre
      style="background: rgb(255, 255, 255); color: rgb(56, 56, 56); font-family: "Fira Code", "Fira Mono", Menlo, Consolas, "DejaVu Sans Mono", monospace; direction: ltr; text-align: left; white-space: pre; word-spacing: normal; word-break: normal; line-height: 1.5; tab-size: 2; hyphens: none; padding: 1em; margin: 0px; overflow: auto; border-radius: 0 0 4px 4px; max-height: 270px;"
    >
      <code
        class="language-sql"
        style="background: rgb(255, 255, 255); color: rgb(56, 56, 56); font-family: "Fira Code", "Fira Mono", Menlo, Consolas, "DejaVu Sans Mono", monospace; direction: ltr; text-align: left; white-space: pre; word-spacing: normal; word-break: normal; line-height: 1.5; tab-size: 2; hyphens: none;"
      >
        <span>
          <span
            class="linenumber react-syntax-highlighter-line-number"
            style="display: inline-block; min-width: 1.25em; padding-right: 1em; text-align: right; user-select: none; font-style: italic; color: rgb(160, 160, 160);"
          >
            1
          </span>
          <span
            class="token"
            style="color: rgb(38, 38, 38);"
          >
            SELECT
          </span>
          <span>
             
          </span>
          <span
            class="token"
            style="color: rgb(64, 64, 64);"
          >
            *
          </span>
          <span>
             
          </span>
          <span
            class="token"
            style="color: rgb(38, 38, 38);"
          >
            FROM
          </span>
          <span>
             
          </span>
          <span
            class="token"
            style="color: rgb(38, 38, 38);"
          >
            table
          </span>
          <span
            class="token"
            style="color: rgb(56, 56, 56);"
          >
            ;
          </span>
        </span>
      </code>
    </pre>
  </div>
</div>
`;

exports[`Test Chat Markdown Should be a memoized component: code block with convert 1`] = `
<div
  class="dtc__aigc__markdown dtc__aigc__markdown--blink jest"
>
  <div
    class="dtc__aigc__codeblock dtc__aigc__codeblock--convert"
  >
    <div
      class="dtc__aigc__codeblock__header"
    >
      <span
        class="dtc__aigc__codeblock__language"
      >
        sql
      </span>
      <div
        class="dtc__aigc__codeblock__tool"
      >
        <span
          class="dtc-copy"
        >
          <span
            class="dtc-copy__default-icon"
            data-mock-icon="CopyOutlined"
          />
        </span>
      </div>
    </div>
    <pre
      style="background: rgb(249, 249, 250); color: rgb(56, 56, 56); font-family: "Fira Code", "Fira Mono", Menlo, Consolas, "DejaVu Sans Mono", monospace; direction: ltr; text-align: left; white-space: pre; word-spacing: normal; word-break: normal; line-height: 1.5; tab-size: 2; hyphens: none; padding: 1em; margin: 0px; overflow: auto; border-radius: 0 0 4px 4px; max-height: 270px;"
    >
      <code
        class="language-sql"
        style="background: rgb(249, 249, 250); color: rgb(56, 56, 56); font-family: "Fira Code", "Fira Mono", Menlo, Consolas, "DejaVu Sans Mono", monospace; direction: ltr; text-align: left; white-space: pre; word-spacing: normal; word-break: normal; line-height: 1.5; tab-size: 2; hyphens: none;"
      >
        <span>
          <span
            class="linenumber react-syntax-highlighter-line-number"
            style="display: inline-block; min-width: 1.25em; padding-right: 1em; text-align: right; user-select: none; font-style: italic; color: rgb(160, 160, 160);"
          >
            1
          </span>
          <span
            class="token"
            style="color: rgb(38, 38, 38);"
          >
            SELECT
          </span>
          <span>
             
          </span>
          <span
            class="token"
            style="color: rgb(64, 64, 64);"
          >
            *
          </span>
          <span>
             
          </span>
          <span
            class="token"
            style="color: rgb(38, 38, 38);"
          >
            FROM
          </span>
          <span>
             
          </span>
          <span
            class="token"
            style="color: rgb(38, 38, 38);"
          >
            table
          </span>
          <span
            class="token"
            style="color: rgb(56, 56, 56);"
          >
            ;
          </span>
        </span>
      </code>
    </pre>
  </div>
</div>
`;
